inverse Laplace transform of exp is shifted Dirac delta

cbm755 opened the pull request:
A constant times an exponential should be a DiracDelta.  Rephrase a
comment to clarify.  Add tests.  This may not be the ideal place for
general improvements to inverse Laplace transforms but if it works...

Fixes #8227.    Not sure there are other relevant issues, but its also
discussed at https://groups.google.com/forum/#!topic/sympy/2FVQRtZFpeY

There is a hack in the second commit to support `exp(s)`: the `Mul` code can handle `exp` if we could just write it has `1*exp`.  I don't see a way to do that, but I can do `2*1/2*exp`.

#### Other comments

I know @asmeurer has suggested that efforts to improve transforms should go into `manualintegrate`...  But this change seems appropriate (to me) within `meijerint_inversion` (which isn't really a user-facing function anyway).

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* integrals
  * inverse Laplace transform of exp is shifted Dirac delta
<!-- END RELEASE NOTES -->

cbm755 committed:
inverse Laplace transform of exp is shifted Dirac delta

A constant times an exponential should be a DiracDelta.  Rephrase a
comment to clarify.  Add tests.  This may not be the ideal place for
general improvements to inverse Laplace transforms but if it works...

Not sure there is a relevant issue number, but discussed a bit here:
https://groups.google.com/forum/#!topic/sympy/2FVQRtZFpeY

cbm755 committed:
inverse laplace transform: fix exp

The `Mul` code can handle `exp` if we could just write it has `1*exp`.
I don't see a way to do that, but I can do `2*1/2*exp`.  Add test.

jksuom commented on sympy/integrals/meijerint.py:
```diff
@@ -2051,9 +2051,15 @@ def meijerint_inversion(f, x, t):
+    if f.func == exp:
+        # exp would work if written as a Mul
+        f = Mul(2, S.Half, f, evaluate=False)
+
```
I would rather try to do something like this:

    if f.is_Mul:
        args = list(f.args)
    elif isinstance(f, exp):
        args = [f]
    else:
        args = None

    if args:
        newargs = []
        ...

cbm755 committed:
meijerint_inversion: simpler logic in shift processing

Instead of forcing `exp` to look like a `Mul`, just build the list
differently.

asmeurer commented:
I suggested manualintegrate because it's easier to extend than meijerint, and fits well for large table lookups that are generally used for transforms (once we have Rubi working we could look at using it instead). If meijerint can be fixed, that's even better.

asmeurer commented:
I couldn't find any related issues. I did find https://github.com/sympy/sympy/issues/2803 but it still gives the same result in this branch, so I guess it's a separate issue.

cbm755 commented:
Thanks for the link!  In fact, this PR fixes #8227.

jksuom commented:
I think that a function can be a Laplace transform only if it is bounded on vertical lines. For an exponential, this means that the shift has to be real. Otherwise there is no inverse transform.

cbm755 commented on sympy/integrals/meijerint.py:
```diff
@@ -2091,6 +2099,11 @@ def meijerint_inversion(f, x, t):
         shift = Add(*exponentials)
         f = Mul(*newargs)
 
+    if f.is_number:
```
```diff
- if f.is_number:
+ if x not in f.free_symbols and shift.is_real:-
```

@jksuom maybe this would be better?  `shift.is_real` is to address your boundedness comment.   With this, we could have:
```
>>> a, b = symbols('a, b', real=True)
>>> meijerint.meijerint_inversion(a*exp(b*s), s, t)
a*DiracDelta(b + t)
```

asmeurer replied:
It would be better to add it as one of the convergence conditions that are returned from the transform function if that's possible.

jksuom replied:
Perhaps something like this could be returned when `f` is constant: 

    Piecewise(f*DiracDelta(t_ + shift), shift.is real), (None, True))

I think that `None` would be better than an unevaluated integral as it would show that the transform does not exist.

asmeurer replied:
You mean `nan`?

jksuom replied:
I thought that `None` would better describe the nonexistence.

asmeurer replied:
But None isn't a symbolic object in SymPy.

Maybe we should create some kind of non-convergent integral object for cases like this (we could also have non-convergent Sum).

cbm755 replied:
I investigated a bit further.

1.  I don't think you're allowed to pass `None` as the cond in a Piecewise.  (And of course `z.is_real` is `None` for general symbols `z`.)

2.  The closest I could find to a function form of `.is_real` is `Contains(shift, Reals)`.  But these don't interact well with `Piecewise`:

```
>>> r = Symbol('r', real=True)
>>> z = Symbol('z')

>>> Piecewise((6, Contains(r, Reals)), (7, True))    # here Contains evaluates to True
6
```
But
```
>>> Piecewise((6, Contains(r, Reals, evaluate=False)), (7, True))
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-10-a59a6e101eb5> in <module>()
----> 1 Piecewise((6, Contains(r, Reals, evaluate=False)), (7, True))

/home/cbm/src/sympy.git/sympy/functions/elementary/piecewise.pyc in __new__(cls, *args, **options)
    134 
    135         if options.pop('evaluate', True):
--> 136             r = cls.eval(*newargs)
    137         else:
    138             r = None

/home/cbm/src/sympy.git/sympy/functions/elementary/piecewise.pyc in eval(cls, *_args)
    183                     x = free.pop()
    184                     try:
--> 185                         c = c.as_set().as_relational(x)
    186                     except NotImplementedError:
    187                         pass

AttributeError: 'Contains' object has no attribute 'as_relational'
```
And thus also
```
>>> Piecewise((6, Contains(z, Reals)), (7, True))
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-12-c87274518aa0> in <module>()
----> 1 Piecewise((6, Contains(z, Reals)), (7, True))

/home/cbm/src/sympy.git/sympy/functions/elementary/piecewise.pyc in __new__(cls, *args, **options)
    134 
    135         if options.pop('evaluate', True):
--> 136             r = cls.eval(*newargs)
    137         else:
    138             r = None

/home/cbm/src/sympy.git/sympy/functions/elementary/piecewise.pyc in eval(cls, *_args)
    183                     x = free.pop()
    184                     try:
--> 185                         c = c.as_set().as_relational(x)
    186                     except NotImplementedError:
    187                         pass

AttributeError: 'Contains' object has no attribute 'as_relational'
```

asmeurer replied:
Contains should work as a boolean, but regardless, the better way to do what you want is `Eq(im(r), 0)`.

asmeurer replied:
I opened https://github.com/sympy/sympy/issues/14977 for `Contains`. It should be easy to fix.

cbm755 replied:
thanks!  I'll revise with `Eq(im(shift), 0)`

jksuom replied:
I am not sure that an integral object would be a good solution. It seems that returning `None` in case the inverse does not exist would suffice. [`_inverse_laplace_transform`](https://github.com/sympy/sympy/blob/master/sympy/integrals/transforms.py#L1164) would then do the right thing, raise `IntegralTransformError`. Otherwise we could probably return a Piecewise with one component. Only the first one is inspected by the caller. So, maybe something like this could work.

    if shift.is_real != False:
        return Piecewise((f*DiracDelta(t_ + shift), shift.is real))
    else:
        return None

If `shift.is_real == True`, it seems that Piecewise will automatically simplify.

cbm755 committed:
meijerint_inversion: allow general constant real shifts

Be more careful about what is returned depending on whether
the shift is real, nonreal or indeterminant.

cbm755 commented on sympy/integrals/tests/test_transforms.py:
```diff
@@ -583,6 +583,21 @@ def simp_hyp(expr):
+    assert ILT(exp(2*s), s, t) == DiracDelta(t + 2)
+    r = Symbol('r', real=True)
+    assert ILT(exp(r*s), s, t) == DiracDelta(t + r)
+    # TODO: what if r.is_real is None/False?  See meijerint_inversion
```
Note the transform code which calls `meijerint_inversion` may not make great choices but I think what we return here is consistent with what the G function part of `meijerint_inversion`.

cbm755 replied:
`inverse_laplace_transform` does the right thing when `noconds=False` is passed.  I've added tests for this.

IMHO, it might be better if `inverse_laplace_transform` just gave Piecewise.  But that's a separate issue.

asmeurer replied:
> IMHO, it might be better if inverse_laplace_transform just gave Piecewise. But that's a separate issue.

+1

cbm755 committed:
meijerint_inversion: shift should be a sympy object

jksuom commented on sympy/integrals/meijerint.py:
```diff
@@ -2091,6 +2099,18 @@ def meijerint_inversion(f, x, t):
+        res = f*DiracDelta(t + shift)
+        _debug('Result is a delta function, possibly conditional:', res, cond)
+        return Piecewise((res.subs(t, t_), cond),
+                         (InverseLaplaceTransform(f_.subs(t, t_), x, t_, None), True))
```
I don't think this (nonexistent) transform would be useful. It seems to be possible to have a Piecewise with only one component.

cbm755 replied:
ok, I've done this.

cbm755 committed:
Merge remote-tracking branch 'origin/master' into invlaplace_exp

cbm755 committed:
meijerint_inversion: simplify logic due to bug fix

Issue #14978 was fixed so we can directly test `cond`.

cbm755 committed:
inverse_laplace_transform: exp test with real/nonreal param

By using `noconds=False` we can test some of the cases in the previous
TODO note.

cbm755 committed:
meijerint_inversion: exp: don't return unevaluated in Piecewise

If `cond` is True, it would be discarded anyway.  If `cond` is an
expression (an Equality) then what we really want to say is that
the transform does not exist.  We don't currently have a way to do
that precisely but a Piecewise we no default `True` subfunction
should be mathematically correct (in that the result is undefined
when the condition is false).  Leave a comment in the code to note
these are the only two options for `cond` (in contrast with the
later code in this function).

jksuom commented on sympy/integrals/meijerint.py:
```diff
@@ -2091,6 +2099,18 @@ def meijerint_inversion(f, x, t):
 
+    if x not in f.free_symbols:
+        _debug('Expression consists of constant and exp shift:', f, shift)
+        from sympy import InverseLaplaceTransform, Eq, im
```
InverseLaplaceTransform no more needed?

cbm755 committed:
meijerint: remove unused import

jksuom commented:
This looks good to me. Maybe something similar could be done in case `f` is not constant, but that would be another task

cbm755 commented:
Thank you both for your help!

> Maybe something similar could be done in case f is not constant

The `shift` part already is used later (in fact, that's why I thought this function was an appropriate place to do this change).
